strawberry/cli/commands/codegen.py

```diff
@@ -143,12 +178,50 @@ def codegen(
     ),
     cli_plugin: Optional[str] = None,
 ) -> None:
+    if not query:
+        return
+
     schema_symbol = load_schema(schema, app_dir)
 
-    console_plugin = _load_plugin(cli_plugin) if cli_plugin else ConsolePlugin
+    console_plugin_type = _load_plugin(cli_plugin) if cli_plugin else ConsolePlugin
 
-    plugins = _load_plugins(selected_plugins)
-    plugins.append(console_plugin(query, output_dir, plugins))
+    plugins = _load_plugins(selected_plugins, query[0])
+    console_plugin = console_plugin_type(query[0], output_dir, plugins)
+
+    if not isinstance(console_plugin, ConsolePlugin):
+        warnings.warn(
+            "The ConsolePlugin should inherit from ``{__name__}.ConsolePlugin``.",
+            DeprecationWarning,
+            stacklevel=1,
+        )
 
-    code_generator = QueryCodegen(schema_symbol, plugins=plugins)
-    code_generator.run(query.read_text())
+    with contextlib.suppress(AttributeError):
+        # This method was not part of the original implementation of the ConsolePlugin
+        # and the ConsolePlugin is overridable on the CLI.  It isn't guaranteed that
+        # custom ConsolePlugin implementations inherit from ``ConsolePlugin``, so if
+        # the implementor does not have this method, we will just keep going.
+        console_plugin.before_any_start()
+
+    for q in query:
+        plugins = _load_plugins(selected_plugins, q)
+        console_plugin.query = q  # update the query in the console plugin.
+        plugins.append(console_plugin)
+
+        for p in plugins:
+            # Adjust the names of the output files for the buildin plugins if there
+            # are multiple files being generated.
+            if len(query) > 1:
+                if isinstance(p, PythonPlugin):
+                    p.outfile_name = q.stem + ".py"
+                elif isinstance(p, TypeScriptPlugin):
+                    p.outfile_name = q.stem + ".ts"
```

Comment:
I wonder if in future we should give the output name to the plugins

Reply:
I'd love that ... I guess I'll raise the question of compatibility and how much we care about making breaking changes in these APIs.

On my side, I'm quite happy to break the API to make it a little cleaner for this direction, but I don't have any idea how impactful that would be for downstream consumers (I don't know how much use this feature actually gets outside of my org 😅 ).

Reply:
those changes are quite simple, so if we provide a guide it should be ok!

Reply:
Excellent.  If you aren't worried about breaking changes then I'll get started on those in the next day or two.

Thanks!

Reply:
uhm, that's a good point!

I guess we can allow plugins to return a list of files, but they should still be able to name them, right?

maybe we can take inspiration from https://the-guild.dev/graphql/codegen ?

Reply:
I'm not exactly sure how that thing works, but I _think_ it's similar to what I'm doing in my custom plugin.

We're using jinja templates to include fragments into the query -- when the codegen sees a named fragment, it will emit an object for that fragment and then that same object can be shared across multiple queries.

(currently, that object gets re-generated with each query that includes it -- it feels sloppy, but since the object generation is deterministic, it works out Ok).

In this paradigm, the plugin is choosing names for the files that include the types that describe the fragments as well as the file that includes the other query specific things.

Reply:
could we give the plugin the source file name and let it decide what to do? 😊

Reply:
> could we give the plugin the source file name and let it decide what to do? 😊

That's what we're doing ... except that the plugin doesn't know how many files they are going to process.  They're designed to process one file at a time.  It could _always_ get the output filename and then decide what to do, but then we don't have enough information to play the "If there is only one file, we'll name the output `types.py`, if there are more than one file, we'll name the output `<stem>.py`".

I'm happy to get rid of the default `types.py` output filename and try to update the corresponding documentation to let people know that the filename will be `<stem>.py` from here on -- is that the direction you'd rather take it?

Reply:
give me some time to think about this properly, I'm clearly tired today :D